ambry-network/src/main/java/com/github/ambry/network/http2/Http2NetworkClient.java

```diff
@@ -79,34 +83,61 @@ public Http2NetworkClient(Http2ClientMetrics http2ClientMetrics, Http2ClientConf
+
+      RequestOrResponse request = (RequestOrResponse) (requestInfo.getRequest());
+      long waitingTime = streamInitiateTime - request.requestCreateTime;
+      http2ClientMetrics.requestToNetworkClientLatencyMs.update(waitingTime);
+
       this.pools.get(InetSocketAddress.createUnresolved(requestInfo.getHost(), requestInfo.getPort().getPort()))
           .acquire()
           .addListener((GenericFutureListener<Future<Channel>>) future -> {
             if (future.isSuccess()) {
               http2ClientMetrics.http2StreamAcquireTime.update(System.currentTimeMillis() - streamInitiateTime);
               long streamAcquiredTime = System.currentTimeMillis();
               Channel streamChannel = future.getNow();
+              correlationIdInFlightToChannelMap.put(requestInfo.getRequest().getCorrelationId(), streamChannel);
               streamChannel.attr(REQUEST_INFO).set(requestInfo);
+              if (!streamChannel.isWritable() || !streamChannel.parent().isWritable()) {
+                http2ClientMetrics.http2StreamNotWritableCount.inc();
+                logger.warn("Stream {} {} not writable. BytesBeforeWritable {} {}", streamChannel.hashCode(),
+                    streamChannel, streamChannel.bytesBeforeWritable(), streamChannel.parent().bytesBeforeWritable());
+              }
               streamChannel.writeAndFlush(requestInfo.getRequest()).addListener(new ChannelFutureListener() {
                 @Override
                 public void operationComplete(ChannelFuture future) throws Exception {
-                  // Listener will be notified after data is removed from ChannelOutboundBuffer (netty's send buffer)
+                  // Listener will be notified right after data is removed from ChannelOutboundBuffer (netty's send buffer)
                   // After removing from ChannelOutboundBuffer, it goes to OS send buffer.
                   if (future.isSuccess()) {
-                    http2ClientMetrics.http2StreamWriteAndFlushTime.update(
-                        System.currentTimeMillis() - streamAcquiredTime);
+                    long writeAndFlushUsedTime = System.currentTimeMillis() - streamAcquiredTime;
+                    http2ClientMetrics.http2StreamWriteAndFlushTime.update(writeAndFlushUsedTime);
                     requestInfo.setStreamSendTime(System.currentTimeMillis());
+                    if (writeAndFlushUsedTime > http2ClientConfig.http2WriteAndFlushTimeoutMs) {
+                      // This usually happens if remote can't accept data in time.
+                      logger.warn(
+                          "WriteAndFlush exceeds http2RequestTimeoutMs {}ms, used time: {}ms, stream channel {}",
+                          http2ClientConfig.http2WriteAndFlushTimeoutMs, writeAndFlushUsedTime, streamChannel);
+                      if (http2ClientConfig.http2DropRequestOnWriteAndFlushTimeout) {
+                        releaseAndCloseStreamChannel(streamChannel);
```

Comment:
Is there a reason for dropping the connection at this point as opposed to just waiting until the router tells the NetworkClient to drop the request after routerRequestTimeoutMs? At this point, the request has already been flushed successfully so the request memory would have already been freed.

Reply:
I agree with you . Changed http2DropRequestOnWriteAndFlushTimeout to false by default.

Reply:
Want to discuss with your guys @justinlin-linkedin @cgtz : 

If a request is flushed successfully(here), its refCnt is decreased, but there is a reference chain that ```AttributeKey<RequestInfo> -> PutRequest -> Bytebuf```, so even refCnt is 0, data is still in memory and can't be recycled. Is this something we can improve?

Reply:
I am worried about this. Comparing to the small object ByteBuf, the byte content held by ByteBuf is way larger. As long as the byte content can be reclaimed by the memory pool, we can ignore the small ByteBuf object.

Reply:
Not fully understand your comment.
If it's PutRquest and it holds big chunk of data, we should break the link between `PutRequst -> ByteBuf`, for GC to recycle, right?

Reply:
Yes and no. When using Unpooled bytebuf, big chunk of data will rely on gc to release the memory. When using Pooled ByteBuf, big chunk of data will rely on calling `release` method to return the chunk back to the pool. GC doesn't reclaim the data chunk, only the small heap memory of ByteBuf fields.